cps/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -767,10 +786,31 @@ synchronized FlowHead getFirstHead() {
         return heads.firstEntry().getValue();
     }
 
-    void notifyListeners(FlowNode node) {
-        if (listeners != null) {
-            for (GraphListener listener : listeners) {
-                listener.onNewHead(node);
+    void notifyListeners(List<FlowNode> nodes, boolean synchronous) {
+        List<GraphListener> _listeners = synchronous ? synchronousListeners : listeners;
+        if (_listeners != null) {
+            Saveable s = Saveable.NOOP;
+            try {
+                Queue.Executable exec = owner.getExecutable();
+                if (exec instanceof Saveable) {
+                    s = (Saveable) exec;
+                }
+            } catch (IOException x) {
+                LOGGER.log(Level.WARNING, null, x);
```

Comment:
The stack trace will be displayed. There is not much more to say. No particular reason to expect this to be thrown.